### PR TITLE
Redux State: Add preview state reducer and actions

### PR DESCRIPTION
**Pull request by sirbrillig**
Relies on #4140 being merged first.

The `DesignPreview` component can render raw markup to a preview and then apply customizations to that preview. This PR adds a data layer for the preview markup and the customizations.

The `customizations` object and the `previewMarkup` used by the preview are stored in the Redux state tree under the `preview` key. The state also keeps track of the "saved" status of the customizations and a history of all customization changes (enabling an "undo" feature).

When the "save" action is dispatched, a special set of functions are called in `client/state/preview/save-functions/index.js`. One function is called for each key in the `customizations` object that's in the current state. Each save function is passed three arguments and is expected to make whatever API calls are necessary to persist that object's data to WordPress. The arguments are:

1. The dispatcher
2. The customizations object for that key
3. The current site ID

This PR includes save functions for the keys `siteTitle` and `headerImage`.

## Testing

There is still nothing hooked up to activate the `design` layout, so to test this, you'll have to pull some code from another branch. To test, hook the 'customize' button in the my-sites sidebar to activate the new layout:

`git cherry-pick 44ba593`

Then click the button and see if the design preview and design menu become visible. The preview iframe should load a preview of the site.

### CODE REVIEW

File: client/state/preview/reducer.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+		case ActionTypes.PREVIEW_CUSTOMIZATIONS_UNDO:
+			const undoneCustomizations = state.previousCustomizations.length > 0 ? state.previousCustomizations.slice( -1 )[0] : {};
+			return assign( {}, state, {
+				isSaved: false,
+				previousCustomizations: state.previousCustomizations.slice( 0, -1 ),
+				customizations: undoneCustomizations,
+			} );
+		case ActionTypes.PREVIEW_CUSTOMIZATIONS_SAVED:
+			return assign( {}, state, { isSaved: true } );
+	}
```
Comment (gwwar) [inline client/state/preview/reducer.js]: You'll want to also implement the `SERIALIZE` or `DESERIALIZE` actions. Without it we'll persist without any safeguards against data drift. If you're not ready to persist this subtree, go ahead and return initialState for both `SERIALIZE` and `DESERIALIZE`.

See: [redux persistence](https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#data-persistence--2754-)

File: client/lib/wpcom-undocumented/lib/site.js [hunk 2]
```diff
@@ -166,6 +166,20 @@ UndocumentedSite.prototype.removeEmailFollower = function( followerId, callback
 	}, callback );
 };
 
+UndocumentedSite.prototype.setHeaderImage = function( imageData, callback ) {
+	this.wpcom.req.post( {
```
Comment (gwwar) [inline client/lib/wpcom-undocumented/lib/site.js]: I find that it's useful to return the post result, so we support both promise and callback style.
```
return this.wpcom.req.post( { //... } );
```

File: assets/stylesheets/layout/_main.scss [hunk 3]
```diff
@@ -375,10 +376,56 @@ $sidebar-width-min: 228px;
 	}
 }
 
+.focus-design {
+	.wp-primary {
+		opacity: 0.2;
+		pointer-events: none;
```
Comment (gwwar) [inline assets/stylesheets/layout/_main.scss]: Any reason why we're disabling pointer-events here?

It'd also be nice to stick to our [css guidelines](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md#csssass-coding-guidelines) here. This seems closer to being a modifer state for the page.

File: assets/stylesheets/layout/_main.scss [hunk 4]
```diff
@@ -375,10 +376,56 @@ $sidebar-width-min: 228px;
...
+.design-menu {
+	background: lighten( $gray, 30% );
+	position: fixed;
+		top: 0;
+		bottom: 0;
+		left: -272px;
+	width: 272px;
+	overflow: hidden;
+	opacity: 0.8;
+	z-index: 181;
```
Comment (gwwar) [inline assets/stylesheets/layout/_main.scss]: Please add new z-indexes to our mapping. https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md#csssass-coding-guidelines

You may find this extension helpful when finding the correct stacking context: https://github.com/gwwar/z-context

File: client/my-sites/design-menu/index.jsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+		return (
+			<div className="design-menu">
+				<span className="current-site__switch-sites">
+					<Button compact borderless onClick={ this.onBack }>
+						<Gridicon icon="arrow-left" size={ 18 } />
+						{ this.translate( 'Back' ) }
+					</Button>
+				</span>
+				<Card>
+					Design Tools!
```
Comment (gwwar) [inline client/my-sites/design-menu/index.jsx]: i18n translation?

File: client/state/preview/reducer.js (hunk 1 above)
Comment (sirbrillig) [inline client/state/preview/reducer.js]: This is *so* good to know. I was hoping there was something like that.

**Comment by sirbrillig** [comment]
Setting to in-progress as some things have changed (current active development is in #4429)

File: assets/stylesheets/layout/_main.scss (hunk 4 above)
Comment (sirbrillig) [inline assets/stylesheets/layout/_main.scss]: Good catch! Those were left-over from the experimental branch.

File: assets/stylesheets/layout/_main.scss (hunk 3 above)
Comment (sirbrillig) [inline assets/stylesheets/layout/_main.scss]: I'll defer to @mtias for those decisions.

File: client/lib/wpcom-undocumented/lib/site.js (hunk 2 above)
Comment (sirbrillig) [inline client/lib/wpcom-undocumented/lib/site.js]: Excellent idea. I also recently learned that there may be another endpoint to set header images which I may use in place of this if I can find it...

**Pull request closed without merging**